pkg/controller/buildrun/runtime_image.go

```diff
@@ -0,0 +1,204 @@
+func splitPaths(dir string) []string {
+	parts := strings.Split(dir, ":")
+	if len(parts) == 2 {
+		return parts
+	}
+	return []string{dir, dir}
+}
+
+// renderEntrypoint will take a slice of strings and render the notation expected on ENTRYPOINT.
+func renderEntrypoint(e []string) string {
+	entrypoint := []string{}
+	for _, cmd := range e {
+		entrypoint = append(entrypoint, strconv.Quote(cmd))
+	}
+	return strings.Join(entrypoint, ", ")
+}
+
+// renderRuntimeDockerfile render runtime Dockerfile using build instance and pre-defined template.
+func renderRuntimeDockerfile(b *buildv1alpha1.Build) (*bytes.Buffer, error) {
+	tmpl, err := template.New(runtimeDockerfile).
+		Funcs(template.FuncMap{
+			"renderUserAndGroup": renderUserAndGroup,
+			"splitPaths":         splitPaths,
+			"renderEntrypoint":   renderEntrypoint,
+		}).
+		Parse(runtimeDockerfileTmpl)
+	if err != nil {
+		return nil, err
+	}
+
+	dockerfile := new(bytes.Buffer)
+	if err = tmpl.Execute(dockerfile, b); err != nil {
+		return nil, err
+	}
+	return dockerfile, nil
+}
+
+// runtimeDockerfileStep trigger the rendering of Dockerfile.runtime, and use this input as a
+// build-step to create a new file.
+func runtimeDockerfileStep(b *buildv1alpha1.Build) (*v1beta1.Step, error) {
+	dockerfile, err := renderRuntimeDockerfile(b)
+	if err != nil {
+		return nil, err
+	}
+
+	container := v1.Container{
+		Name:  "runtime-dockerfile",
+		Image: b.Spec.BuilderImage.ImageURL,
+		SecurityContext: &v1.SecurityContext{
+			RunAsUser: &rootUserID,
+		},
+		WorkingDir: workspaceDir,
+		Command:    []string{"/bin/bash"},
+		Args: []string{
+			"-x",
+			"-c",
+			fmt.Sprintf("echo '%s' >%s", dockerfile, runtimeDockerfilePath),
+		},
+	}
+	return &v1beta1.Step{Container: container}, nil
+}
+
+// runtimeBuildAndPushStep returns a Task step to build the Dockerfile.runtime with kaniko.
+func runtimeBuildAndPushStep(b *buildv1alpha1.Build, kanikoImage string) *v1beta1.Step {
+	contextDir := workspaceDir
+	if b != nil && b.Spec.Source.ContextDir != nil {
+		contextDir = path.Join(workspaceDir, *b.Spec.Source.ContextDir)
+	}
+
+	container := v1.Container{
+		Name:       "kaniko-build-and-push",
+		Image:      kanikoImage,
+		WorkingDir: workspaceDir,
+		SecurityContext: &v1.SecurityContext{
+			RunAsUser: &rootUserID,
+			Capabilities: &v1.Capabilities{
+				Add: []v1.Capability{
+					v1.Capability("CHOWN"),
+					v1.Capability("DAC_OVERRIDE"),
+					v1.Capability("FOWNER"),
+					v1.Capability("SETGID"),
+					v1.Capability("SETUID"),
+				},
+			},
+		},
+		Env: []v1.EnvVar{
+			{Name: "DOCKER_CONFIG", Value: "/tekton/home/.docker"},
+			{Name: "AWS_ACCESS_KEY_ID", Value: "NOT_SET"},
+			{Name: "AWS_SECRET_KEY", Value: "NOT_SET"},
+		},
+		Command: []string{"/kaniko/executor"},
+		Args: []string{
+			"--skip-tls-verify=true",
+			fmt.Sprintf("--dockerfile=%s", runtimeDockerfile),
+			fmt.Sprintf("--context=%x", contextDir),
+			fmt.Sprintf("--destination=%s", b.Spec.Output.ImageURL),
+		},
```

Comment:
Please add `--snapshotMode=redo` like I do in https://github.com/redhat-developer/build/pull/286.

Reply:
Done.